vim-patch:8.2.{1967,2058,2467}

seandewar opened the pull request:
One small difference for 8.2.2058: the `au BufWinEnter * silent loadview` line in `Test_mkview_loadview_jumplist()` was moved down a few lines until after the views are written, as `:loadview` in Nvim will throw E484 if it can't open the view (and also to avoid #13490).

Hopefully this is fine. 😄

seandewar committed:
vim-patch:8.2.1967: the session file does not restore the alternate file

Problem:    The session file does not restore the alternate file.
Solution:   Add ":balt".  Works like ":badd" and also sets the buffer as the
            alternate file.  Use it in the session file. (closes vim/vim#7269,
            closes vim/vim#6714)
https://github.com/vim/vim/commit/59d8e56e048eb5d384649284fb35363931fc3697

Include minimal test_buffer.vim from patch 8.2.0243 for Test_balt().
Add entry for :balt to runtime/doc/index.txt from vim/vim#7819.

seandewar committed:
vim-patch:8.2.2058: using mkview/loadview changes the jumplist

Problem:    Using mkview/loadview changes the jumplist.
Solution:   Use ":keepjumps".  Don't let ":badd" or ":balt" change the
            jumplist. (closes vim/vim#7371)
https://github.com/vim/vim/commit/3482be6a33c3a00fe1e8e604310ad3034f593045

Definition for BufWinEnter autocmd in Test_mkview_loadview_jumplist()
was moved down a few lines until after the views are written, as in
Nvim, :loadview throws if the view file can't be opened.

janlazo commented:
Going to review this after resolving some of the patches in https://github.com/neovim/neovim/pull/13079. You added a test from `test_buffer.vim` but `test_buffer.vim` is not merged yet.

seandewar commented:
Thanks! Be sure to let me know if anything needs changing. :+1:
After 8.1.1241 is ported I'll be able to use `ADDR_NONE` over `ADDR_LINES` for `:balt`.

> You added a test from test_buffer.vim but test_buffer.vim is not merged yet.

Sorry, forgot to mention that here; it's originally from 8.2.0243. There's a small comment about it in the commit message.

markwu commented:
Will this PR merge to neovim? WIthout `:balt` There is no way to load vim saved session.

janlazo commented:
Patch v8.1.1241 was merged in https://github.com/neovim/neovim/pull/13655

seandewar commented:
Noted, cheers! I'll try to revisit this PR tomorrow when I have some time free. :smile:

seandewar commented:
Sorry about the delay; been a busy few weeks!

I've re-checked things and updated the `ex_cmds.lua` entry to reflect #13655.
Let me know if I've missed anything.

janlazo commented:
It's risky to include this in the 0.5.0 release because tests written after https://github.com/vim/vim/commit/9b7bf9e98f06ece595fed7a3ff53ecce89797a53 may expect different errors.

janlazo commented:
Check if https://github.com/vim/vim/commit/139348f3e8370826a1ceb10c73f7c7bb586f8125 can be included here.

seandewar committed:
vim-patch:8.2.2467: script generated by :mkview changes alternate file

Problem:    Script generated by :mkview changes alternate file.
Solution:   Only write :balt in the session file. (Harish Rajagopal,
            closes vim/vim#7779)
https://github.com/vim/vim/commit/139348f3e8370826a1ceb10c73f7c7bb586f8125

janlazo commented on runtime/doc/index.txt:
```diff
@@ -1144,6 +1144,7 @@ tag		command		action ~
 |:bNext|	:bN[ext]	go to previous buffer in the buffer list
 |:ball|		:ba[ll]		open a window for each buffer in the buffer list
 |:badd|		:bad[d]		add buffer to the buffer list
+|:balt|		:balt		like ":badd" and set the alternate file
```
This is missing in Vim. Keep it anyway.

seandewar replied:
Vim will soon have this as of vim/vim#7819. :+1:

janlazo commented on src/nvim/ex_cmds.h:
```diff
@@ -16,7 +16,8 @@
 #define ECMD_FORCEIT         0x08    // ! used in Ex command
 #define ECMD_ADDBUF          0x10    // don't edit, just add to buffer list
-
+#define ECMD_ALTBUF          0x20    // like ECMD_ADDBUF and set the alternate
```
This can be reindented instead so that it fits in 1 line like Vim. Neovim doesn't use tabs and avoids vertical alignment to avoid whitespace changes in some parts of the codebase (ie. src/nvim/eval/ function docstrings).

janlazo commented on src/nvim/testdir/test_mksession.vim:
```diff
@@ -399,6 +413,54 @@ func Test_mkview_no_file_name()
+  call assert_equal([3, 4], getcurpos()[1:2])
+  write
+
+  au BufWinEnter * silent loadview
```
This is what you moved from the line after `au BufWinLeave * silent mkview`. Leave a comment here to explain why this is needed and restore this line in its original spot but comment it out for reference.

Is this still needed?

seandewar replied:
Still seems so.

Because `:loadview` is triggered on `BufWinEnter`, it'll run on the initial `:edit`s before the views are saved with `:mkview`, which triggers on `BufWinLeave`.

In Vim, `:loadview` silently fails if the view isn't found (doesn't use the return code from `do_source()`).
Nvim checks `do_source()`'s return code and raises `e_notopen` (E484) instead.

The lines above are only to create the views, `:loadview` is only needed after the views exist.
I'll include the original line as a comment and add the reasoning. :+1:

Also, on another note, #13490 still happens if the `:loadview` errors inside the `:au`.
It'll still cause the test to fail (from the leak + ASAN) if this were to happen, but that probably isn't ideal. 😄

janlazo commented:
I'm ok with merging this in v0.5.0 as long as test changes are documented in the commit message and within the test via comments.
